File: haskell/glue/src/Glue/Lib/List/Cdr.hs
module Glue.Lib.List.Cdr where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

cdr :: IR Eval
cdr = NativeFunc cdrImpl

cdrImpl :: IR Eval -> Eval (IR Eval)
cdrImpl arg = case arg of
    List (_ : xs) -> pure $ List xs
    List [] -> throwError $ wrongArgumentType ["non-empty list"]
    _ -> throwError $ wrongArgumentType ["list"]

File: dart/glue/lib/src/lib/list/cdr.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Cdr function - returns the rest of a list after the first element
/// Mirrors Haskell Glue.Lib.List.Cdr.cdr exactly
Ir cdr = IrNativeFunc(cdrImpl);

/// Cdr function implementation
/// Mirrors Haskell Glue.Lib.List.Cdr.cdrImpl exactly
Eval<Ir> cdrImpl(Ir arg) {
  return switch (arg) {
    IrList(:final elements) =>
      elements.isEmpty
          ? throwError(wrongArgumentType(['non-empty list']))
          : Eval.pure(IrList(elements.skip(1).toList())),
    _ => throwError(wrongArgumentType(['list'])),
  };
}
